mqs/jwks_auth.py

```diff
@@ -0,0 +1,157 @@
+            algorithm=BROKER_QUEUE_AUTH_ALGO,
+        )
+
+        return jwt_auth_handler.create_token(
+            "mqs",
+            expiration=config.ENV.BROKER_QUEUE_AUTH_TOKEN_EXP,
+            payload={
+                # https://www.rabbitmq.com/docs/oauth2#prerequisites
+                # matches rabbitmq broker's 'resource_server_id' value
+                "aud": config.ENV.BROKER_RESOURCE_SERVER_ID,
+                # https://www.rabbitmq.com/docs/oauth2#rich-authorization-request
+                "authorization_details": [
+                    {
+                        # https://www.rabbitmq.com/docs/oauth2#type-field
+                        # -> matches rabbitmq broker's 'resource_server_type' value
+                        "type": "rabbitmq",
+                        # https://www.rabbitmq.com/docs/oauth2#locations-field
+                        "locations": [
+                            f"cluster:^{config.ENV.BROKER_RESOURCE_SERVER_ID}$/queue:^{mqid}$"
+                        ],
+                        # https://www.rabbitmq.com/docs/oauth2#actions-field
+                        "actions": ["read", "write"],
+                    }
+                ],
+            },
+            headers={"kid": self.kid},
+        )
+
+    async def _update_jwks_in_db(self) -> None:
+        """Add a JWK to the database and remove old ones.
+
+        Triggered when a public key is updated.
+        """
+        # check if this was actually an update; else, the process just restarted
+        if await self._mongo_collection.find_one({"kid": self.kid}):
+            LOGGER.info("Updated broker-queue auth public key is already in db")
+            return
+
+        # set exp on "current" jwk
+        await self._mongo_collection.update_many(  # expected to be only 1 doc
+            {
+                "_exp": None,  # get those w/ field that is None or doesn't exist
+            },
+            {
+                # set exp so it is greater than the last jwt generated using jwk's pub key
+                "$set": {"_exp": time.time() + config.ENV.BROKER_QUEUE_AUTH_TOKEN_EXP},
+            },
+        )
+        LOGGER.info("Set expiration for previous JWK in db")
+
+        # insert new jwk
+        key_obj = RSAAlgorithm(RSAAlgorithm.SHA256).prepare_key(
+            key=await self.get_public_key()
+        )
+        jwk = {
+            "kid": self.kid,
+            "_exp": None,
+            **RSAAlgorithm.to_jwk(key_obj, as_dict=True),
+        }
+        await self._mongo_collection.insert_one(jwk)
+        LOGGER.info("Added new JWK to db")
+
+        # remove any expired
+        res = await self._mongo_collection.delete_many({"_exp": {"$lt": time.time()}})
+        LOGGER.info(f"Deleted {res.deleted_count} expired JWKs")
+
+    async def get_jwks_from_db(self) -> list[dict[str, str]]:
+        """Retrieve the JWKS list from the database."""
+        jwks = [d async for d in self._mongo_collection.find()]
```

Comment:
If we haven't gotten an update in a while, there may be an expired jwk in the db.  Maybe query for only non-expired jwks here?